articles/views.py

```diff
@@ -5,14 +5,79 @@
 from django.contrib.auth.decorators import login_required
 from django.contrib.auth.mixins import LoginRequiredMixin
 from django.views.generic.edit import CreateView, UpdateView
-from .models import Article
+from django.views.generic import ListView
+from django.db.models import Count
+from .models import Article, Tag
 from .forms import ArticleForm
 
 
 def view(req, slug):
     return render(req, "articles/detail.html")
 
 
+class ListArticle(ListView):
+    template_name = "articles/list.html"
+    paginate_by = 10
+    model = Article
+
+    def _get_query_context(self):
+        ctx = {
+            "tag": None,
+            "query_string": [],
+            "author": None,
+            "favorited": None,
+            "followed_by": None,
+        }
+
+        tag_slug = self.request.GET.get("tag")
+        if tag_slug:
+            tag = get_object_or_404(Tag, slug=tag_slug)
+            ctx["tag"] = tag
+            ctx["query_string"].append(f"tag={tag_slug}")
+
+        author = self.request.GET.get("author")
+        if author:
+            ctx["author"] = author
+            ctx["query_string"].append(f"author={author}")
+
+        favorited = self.request.GET.get("favorited")
+        if favorited:
+            ctx["favorited"] = favorited
+            ctx["query_string"].append(f"favorited={favorited}")
+
+        followed_by = self.request.GET.get("followed_by")
+        if followed_by:
+            ctx["followed_by"] = followed_by
+            ctx["query_string"].append(f"followed_by={followed_by}")
+
+        ctx["query_string"] = "&".join(ctx["query_string"])
+        return ctx
+
+    def get_queryset(self):
+        qs = (
+            Article.objects.all()
+            .select_related("author__user")
+            .annotate(favorited_by__count=Count("favorited_by"))
+            .order_by("-created_at")
+        )
+
+        if tag := self._get_query_context()["tag"]:
+            qs = qs.filter(tags__slug=tag.slug)
+
+        if author := self._get_query_context()["author"]:
+            qs = qs.filter(author__user__username=author)
+
+        if favorited := self._get_query_context()["favorited"]:
+            qs = qs.filter(favorited_by__user__username=favorited)
+
+        return qs
+
+    def get_context_data(self, **kwargs):
+        ctx = super().get_context_data(**kwargs)
+        ctx.update(self._get_query_context())
+        return ctx
```

Comment:
This list view can filter based on a few different query params activated by the parent page's tabs.

Reply:
Could probably add some model managers to Article for the different use cases, especially if we need them in other views.

Reply:
Ah, that's a good idea! We don't need them right now, but would clean up this view a bit. I'm taking over the "favorite" PR right now. I'm going to see if that will help there.